### ISSUE TITLE
Strange expiry time issues

### ISSUE THREAD

**Issue by gavinsykes** [lcobucci/jwt#978]
Hi all 👋🏻  hoping someone can help with a strange issue I'm having!

When I generate an access token with JWT all seems well, but then when I return the access token to my system for validation, it somehow returns that the token expired the day before (in fact, 2 days ago now I'm in the day after having this issue) and I'm wondering what, if anything, I'm doing wrong.

```php
// Token Generation

$now = (new \DateTimeImmutable())->setTimezone(new \DateTimeZone('UTC'));
// I was having this issue before using setTimezone as well so I'm fairly sure it's not that, plus I want everything on my API to be in UTC anyway!

$access_token = $token_builder
    ->issuedBy('my-api')
    ->permittedFor('my-frontend-application')
    ->identifiedBy(uniqid())
    ->issuedAt($now)
    ->canOnlyBeUsedAfter($now)
    ->expiresAt($now->modify('+1 hour'))
    ->withClaim('user_id', $authentication['user_id'])
    ->withClaim('user_roles', $roles)
    ->getToken($algorithm, $signing_key);
```

Other than the custom claims, this is almost exactly wht is in the documentation.

👍🏻 Now if I immediately take that token and log it to my terminal, let's say I created it at 2022-11-16T10:00:00Z, it will give me a DateTimeImmutable object like so:

```php
DateTimeImmutable Object
(
    [date] => 2022-11-16 11:00:00.000000
    [timezone_type] => 3
    [timezone] => UTC
)
```

All good so far.

👍🏻 If I take the result of `$access_token->toString()` and put it into a JWT validator like jwt.io, that will tell me the exp claim as a number which you can hover over to see it corresponds to 2022-11-16T11:00:00Z, still all good.

👎🏻 Now if I take the token string and try to use it to access an API endpoint where there's middleware to validate the token (code below, and I'm using Slim V4 if that matters), I get back a DateTimeImmutable object like the below, and I suppose my main question is, well...how?!

```php
DateTimeImmutable Object
(
    [date] => 2022-11-14 17:31:59.167623
    [timezone_type] => 1 // I've never found much on what the timezone_type is meant to do or signify, but...
    [timezone] => +00:00 // This is the same as UTC so does that really matter?
)
```

```php
$parser = new JWTParser(new JoseEncoder());
$access_token = explode('Bearer ',$request->getHeader('Authorization')[0])[1];
$token = $parser->parse($access_token);
assert($token instanceof UnencryptedToken);
$validator = new JWTValidator();
if (!$validator->validate($token, new IssuedBy('my-api'))) {
  throw new AuthenticationException('The provided token was invalid');
}
if (!$validator->validate($token, new PermittedFor('my-frontend-application'))) {
  throw new AuthenticationException('The provided token was invalid');
}
if (!$validator->validate($token, new StrictValidAt(FrozenClock::FromUTC()))) {
  // I can't find much on the FrozenClock object but it gives a DateTimeImmutable equating to "now" so it's what I need
  log_error("time invalid");
  log_error(print_r(FrozenClock::fromUTC(), true));
  log_error(print_r($token->claims()->get('exp'), true)); // This is what returns the DateTimeImmutable object above
  throw new AuthenticationException('The provided token was invalid');
}
```

I cannot work out if it's something I'm doing wrong or perhaps the parser having issues, but surely if it were the parser then this would be a much wider issue that would have already been reported and maybe even fixed?

**Comment by gavinsykes** [comment]
Update after further testing: it doesn't seem to matter what I actually send as the access token, it interprets the expiry time as 2022-11-14 17:31:59.167623 and nothing else.

**Comment by lcobucci** [comment]
Time objects are always encoded in timestamps, which will always be UTC. PHP's comparison is timezone aware so that should not cause any harm.

I'm curious why you chose to use the frozen clock implementation... Did you check its use case?

**Comment by lcobucci** [comment]
Also, do you have a sample token?

**Comment by gavinsykes** [comment]
Hi @lcobucci I wasn't really familiar with whether to use FrozenClock or SystemClock and while that doesn't appear to be the issue, on your advice I'll use an alternative which I'm guessing would be SystemClock.

Regarding the actual cause of the issue, it seems that an apology is in order; I've discovered what was causing it and it was me being stupid!

I'm using Insomnia to test the API and in my earlier testing I just grabbed an access token and stuck it in the Headers tab to test "does this token get parsed correctly and give me what I need?" Which it did, hooray! However, I then didn't take that access token out of the Headers tab for further testing, meaning that no matter what access token Insomnia got back from the API when fetching a new token, the header would always have that original one! I've just gone now to check "what is it actually sending in the header?" And discovered the original token still sitting there.

I have spent *three whole days* on this issue and it was caused entirely by me being an idiot 🤦🏻‍♂️ 😱 😡 

I am so sorry!
